core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -448,45 +447,84 @@ private RexNode simplifyCase(RexCall call) {
-      //   WHEN p3 THEN TRUE
-      //   ELSE FALSE
-      //   END
-      // if p1...pn cannot be nullable
-      for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
-        if (pair.e.getKey().getType().isNullable()) {
-          break trueFalse;
+      boolean anyNullable = false;
+    trueFalse:
+      if (true) {
+        // 2) Another simplification
+        //   CASE
+        //   WHEN p1 THEN TRUE
+        //   WHEN p2 THEN FALSE
+        //   WHEN p3 THEN TRUE
+        //   ELSE FALSE
+        //   END
+        // if p1...pn cannot be nullable
+        for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
+          if (pair.e.getKey().getType().isNullable()) {
+            anyNullable = true;
+            break trueFalse;
+          }
+          if (!pair.e.getValue().isAlwaysTrue()
+              && !pair.e.getValue().isAlwaysFalse()
+              && (!unknownAsFalse || !RexUtil.isNull(pair.e.getValue()))) {
+            break trueFalse;
+          }
+        }
+        final List<RexNode> terms = new ArrayList<>();
+        final List<RexNode> notTerms = new ArrayList<>();
+        for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
+          if (pair.e.getValue().isAlwaysTrue()) {
+            terms.add(RexUtil.andNot(rexBuilder, pair.e.getKey(), notTerms));
+          } else {
+            notTerms.add(pair.e.getKey());
+          }
         }
-        if (!pair.e.getValue().isAlwaysTrue()
-            && !pair.e.getValue().isAlwaysFalse()
-            && (!unknownAsFalse || !RexUtil.isNull(pair.e.getValue()))) {
-          break trueFalse;
+        final RexNode disjunction = RexUtil.composeDisjunction(rexBuilder, terms);
+        if (!call.getType().equals(disjunction.getType())) {
+          return rexBuilder.makeCast(call.getType(), disjunction);
         }
+        return disjunction;
       }
-      final List<RexNode> terms = new ArrayList<>();
-      final List<RexNode> notTerms = new ArrayList<>();
-      for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
-        if (pair.e.getValue().isAlwaysTrue()) {
-          terms.add(RexUtil.andNot(rexBuilder, pair.e.getKey(), notTerms));
-        } else {
-          notTerms.add(pair.e.getKey());
+      if (!anyNullable) {
+        // 3 Simplify CASE into AND/OR:
+        // CASE
+        // WHEN p1 THEN e1
+        // WHEN p2 THEN e2
+        // ELSE TRUE/FALSE
+        // END
+        // (p1 and e1) or (!p1 and p2 and e2) or (!p1 and !p2 and <else>)
+        // p1..pn not nullable
+
+        RexNode notPredicates = rexBuilder.makeLiteral(true);
+        final List<RexNode> terms = new ArrayList<>();
+        for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
+          final RexNode toAdd = RexUtil.composeConjunction(rexBuilder,
+            ImmutableList.<RexNode>of(
+              notPredicates,
+              pair.e.getKey(),
+              pair.e.getValue()),
+            false);
+          terms.add(simplify(toAdd));
```

Comment:
We might miss some simplification if we call simplify at every iteration vs over the returned disjunction (L514), right? (e.g., folding of OR true)